In [1]:
#  Ebnable HTML/CSS 
from IPython.core.display import HTML
HTML("<link href='https://fonts.googleapis.com/css?family=Passion+One' rel='stylesheet' type='text/css'><style>div.attn { font-family: 'Helvetica Neue'; font-size: 30px; line-height: 40px; color: #FFFFFF; text-align: center; margin: 30px 0; border-width: 10px 0; border-style: solid; border-color: #5AAAAA; padding: 30px 0; background-color: #DDDDFF; }hr { border: 0; background-color: #ffffff; border-top: 1px solid black; }hr.major { border-top: 10px solid #5AAA5A; }hr.minor { border: none; background-color: #ffffff; border-top: 5px dotted #CC3333; }div.bubble { width: 65%; padding: 20px; background: #DDDDDD; border-radius: 15px; margin: 0 auto; font-style: italic; color: #f00; }em { color: #AAA; }div.c1{visibility:hidden;margin:0;height:0;}div.note{color:red;}</style>")

___
Enter Team Member Names here (double click to edit):

- Name 1: Triston Hudgins
- Name 2: Shijo Joseph
- Name 3: Douglas Yip

________

# In Class Assignment Three
In the following assignment you will be asked to fill in python code and derivations for a number of different problems. Please read all instructions carefully and turn in the rendered notebook (or HTML of the rendered notebook)  before the end of class.

<a id="top"></a>
## Contents
* <a href="#Loading">Loading the Data</a>
* <a href="#distance">Measuring Distances</a>
* <a href="#KNN">K-Nearest Neighbors</a>
* <a href="#naive">Naive Bayes</a>

________________________________________________________________________________________________________
<a id="Loading"></a>
<a href="#top">Back to Top</a>
## Downloading the Document Data
Please run the following code to read in the "20 newsgroups" dataset from sklearn's data loading module.

In [2]:
from __future__ import print_function

In [3]:
from sklearn.datasets import fetch_20newsgroups_vectorized
import numpy as np

# this takes about 30 seconds to compute, read the next section while this downloads
ds = fetch_20newsgroups_vectorized(subset='train')

# this holds the continuous feature data (which is tfidf)
print('features shape:', ds.data.shape) # there are ~11000 instances and ~130k features per instance
print('target shape:', ds.target.shape) 
print('range of target:', np.min(ds.target),np.max(ds.target))
print('Data type is', type(ds.data), float(ds.data.nnz)/(ds.data.shape[0]*ds.data.shape[1])*100, '% of the data is non-zero')

features shape: (11314, 130107)
target shape: (11314,)
range of target: 0 19
Data type is <class 'scipy.sparse.csr.csr_matrix'> 0.1214353154362896 % of the data is non-zero


## Understanding the Dataset
Look at the description for the 20 newsgroups dataset at http://qwone.com/~jason/20Newsgroups/. You have just downloaded the "vectorized" version of the dataset, which means all the words inside the articles have gone through a transformation that binned them into 130 thousand features related to the words in them.  

**Question Set 1**:
- How many instances are in the dataset? 
- What does each instance represent? 
- How many classes are in the dataset and what does each class represent?
- Would you expect a classifier trained on this data would generalize to documents written in the past week? Why or why not?
- Is the data represented as a sparse or dense matrix?

How many instances are in the dataset?

> Answer - 11314

What does each instance represent?

> Answer - Each newsgroup document written by the 20 different newsgroups

How many classes are in the dataset and what does each class represent?

> Answer - (range of target: 0 19) There are 20 classes in the dataset and each represents a different newsgroup.

Would you expect a classifier trained on this data would generalize to documents written in the past week? Why or why not?

> Answer - No because the events are from 25 years ago and literature and written has changed over the quarter century.

Is the data represented as a sparse or dense matrix?

> Answer - Dense matrix




___
<a id="distance"></a>
<a href="#top">Back to Top</a>
## Measures of Distance
In the following block of code, we isolate three instances from the dataset. The instance "`a`" is from the group *computer graphics*, "`b`" is from from the group *recreation autos*, and "`c`" is from group *recreation motorcycle*. **Exercise for part 2**: Calculate the: 
- (1) Euclidean distance
- (2) Cosine distance 
- (3) Jaccard similarity 

between each pair of instances using the imported functions below. Remember that the Jaccard similarity is only for binary valued vectors, so convert vectors to binary using a threshold. 

**Question for part 2**: Which distance seems more appropriate to use for this data? **Why**?

In [8]:
from scipy.spatial.distance import cosine
from scipy.spatial.distance import euclidean
from scipy.spatial.distance import jaccard
import numpy as np

# get first instance (comp)
idx = 550
a = ds.data[idx].todense()
a_class = ds.target_names[ds.target[idx]]
print('Instance A is from class', a_class)

# get second instance (autos)
idx = 4000
b = ds.data[idx].todense()
b_class = ds.target_names[ds.target[idx]]
print('Instance B is from class', b_class)

# get third instance (motorcycle)
idx = 7000
c = ds.data[idx].todense()
c_class = ds.target_names[ds.target[idx]]
print('Instance C is from class', c_class)

# Enter distance comparison below for each pair of vectors:
p = 'Placeholder'
print('\n\nEuclidean Distance\n ab:', format(euclidean(a,b), ".3f"), 'ac:', format(euclidean(a,c), ".3f"), 'bc:',format(euclidean(b,c), ".3f"))
print('\nCosine Distance\n ab:', format(cosine(a,b), ".3f"), 'ac:', format(cosine(a,c), ".3f"), 'bc:', format(cosine(b,c), ".3f"))

#determine Jaccard binary for a b and c
a_binary = np.where(a > a.mean(), True, False)
b_binary = np.where(b > b.mean(), True, False)
c_binary = np.where(c > c.mean(), True, False)
print('\nJaccard Dissimilarity (vectors should be boolean values)\n ab:', format(jaccard(a_binary,b_binary), ".3f"), 'ac:', format(jaccard(a_binary,c_binary), ".3f"), 'bc:', format(jaccard(b_binary,c_binary), ".3f"))

print('\n\nEuclidean Distance creates the overall largest distance between classes which creates the most distinction.')


Instance A is from class comp.graphics
Instance B is from class rec.autos
Instance C is from class rec.motorcycles


Euclidean Distance
 ab: 1.099 ac: 1.189 bc: 0.918

Cosine Distance
 ab: 0.603 ac: 0.707 bc: 0.421

Jaccard Dissimilarity (vectors should be boolean values)
 ab: 0.882 ac: 0.875 bc: 0.909


Euclidean Distance creates the overall largest distance between classes which creates the most distinction.


___
# Start of Live Session Assignment
___
<a id="KNN"></a>
<a href="#top">Back to Top</a>
## Using scikit-learn with KNN
Now let's use stratified cross validation with a holdout set to train a KNN model in `scikit-learn`. Use the example below to train a KNN classifier. The documentation for `KNeighborsClassifier` is here: http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html  

**Exercise for part 3**: Use the code below to test what value of `n_neighbors` works best for the given data. *Note: do NOT change the metric to be anything other than `'euclidean'`. Other distance functions are not optimized for the amount of data we are working with.* 

**Question for part 3**: What is the accuracy of the best classifier you can create for this data (by changing only the `n_neighbors` parameter)? 

In [9]:
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from IPython.html import widgets 

cv = StratifiedShuffleSplit(n_splits=1, test_size = 0.5, train_size=0.5)

# fill in the training and testing data and save as separate variables
for trainidx, testidx in cv.split(ds.data,ds.target):
    # note that these are sparse matrices
    X_train = ds.data[trainidx] 
    X_test = ds.data[testidx] 
    y_train = ds.target[trainidx]
    y_test = ds.target[testidx]

# fill in your code  here to train and test
# calculate the accuracy and print it for various values of K
#K = 5 
#clf = KNeighborsClassifier(n_neighbors=K, weights='uniform', metric='euclidean')


#acc = 0.2 # overwrite this with the actual accuracy

#=====================================

#print('Accuracy of classifier with %d neighbors is: %.2f'%(K,acc))


Accuracy of classifier with 5 neighbors is: 0.20


C:\Datascience\Anaconda3\envs\ML7331\lib\site-packages\IPython\html.py:14: ShimWarning: The `IPython.html` package has been deprecated since IPython 4.0. You should import from `notebook` instead. `IPython.html.widgets` has moved to `ipywidgets`.
  "`IPython.html.widgets` has moved to `ipywidgets`.", ShimWarning)


In [10]:
#create loop on the formula from question to identify the best k
results = []
for i in range(1,30):

    k = i
    clf = KNeighborsClassifier(n_neighbors=k, weights='uniform', metric='euclidean')
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
        
    print('\n\033[1m' + 'Accuracy of classifier ->'+ '\033[0m', format(accuracy, ".3f"), '[1m' +"K -"+ '\033[0m', k)
    results.append((accuracy,k))


Accuracy: 0.596 For K: 1

Accuracy: 0.558 For K: 2

Accuracy: 0.527 For K: 3

Accuracy: 0.507 For K: 4

Accuracy: 0.495 For K: 5

Accuracy: 0.476 For K: 6

Accuracy: 0.464 For K: 7

Accuracy: 0.453 For K: 8

Accuracy: 0.446 For K: 9

Accuracy: 0.440 For K: 10

Accuracy: 0.432 For K: 11

Accuracy: 0.426 For K: 12

Accuracy: 0.420 For K: 13

Accuracy: 0.413 For K: 14

Accuracy: 0.409 For K: 15

Accuracy: 0.403 For K: 16

Accuracy: 0.399 For K: 17

Accuracy: 0.393 For K: 18

Accuracy: 0.386 For K: 19

Accuracy: 0.378 For K: 20

Accuracy: 0.377 For K: 21

Accuracy: 0.375 For K: 22

Accuracy: 0.369 For K: 23

Accuracy: 0.363 For K: 24

Accuracy: 0.359 For K: 25

Accuracy: 0.356 For K: 26

Accuracy: 0.352 For K: 27

Accuracy: 0.349 For K: 28

Accuracy: 0.345 For K: 29


The best accuracy is 0.596 with K=1 neighbors.

**Question for part 3**: With sparse data, does the use of a KDTree representation make sense? Why or Why not?

Enter your answer below
> -No, because sparse data sets cannot be divided by null values.
___

_____
## KNN extensions - Centroids
Now lets look at a very closely related classifier to KNN, called nearest centroid. In this classifier (which is more appropriate for big data scenarios and sparse data), the training step is used to calculate the centroids for each class. These centroids are saved. Unknown attributes, at prediction time, only need to have distances calculated for each saved centroid, drastically decreasing the time required for a prediction. 

**Exercise for part 4**: Use the template code below to create a nearest centroid classifier. Test which metric has the best cross validated performance: Euclidean, Cosine, or Manhattan. In `scikit-learn` you can see the documentation for NearestCentroid here: 
- http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestCentroid.html#sklearn.neighbors.NearestCentroid

and for supported distance metrics here:
- http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.distance_metrics.html#sklearn.metrics.pairwise.distance_metrics

In [16]:
from sklearn.neighbors import NearestCentroid

# the parameters for the nearest centroid metric to test are:
#    l1, l2, and cosine (all are optimized)
# fill in your code here

clf = NearestCentroid(metric='euclidean')
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
accuracy_eu = accuracy_score(y_test, y_pred)
print('Euclidian Percentage Accuracy:'+ '\033[0m', format(accuracy_eu, ".3f"))

clf = NearestCentroid(metric='cosine')
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
accuracy_co = accuracy_score(y_test, y_pred)
print('Cosine Percentage Accuracy:'+ '\033[0m', format(accuracy_co, ".3f"))

clf = NearestCentroid(metric='manhattan')
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
accuracy_ma = accuracy_score(y_test, y_pred)
print('Manhattan Percentage Accuracy:'+ '\033[0m', format(accuracy_ma, ".3f"))

print('The best distance metric is: Cosine')

Euclidian Percentage Accuracy: 0.398


C:\Datascience\Anaconda3\envs\ML7331\lib\site-packages\sklearn\neighbors\_nearest_centroid.py:165: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  "Averaging for metrics other than "


Cosine Percentage Accuracy: 0.481
Manhattan Percentage Accuracy: 0.174
The best distance metric is: Cosine


___
<a id="naive"></a>
<a href="#top">Back to Top</a>
## Naive Bayes Classification
Now let's look at the use of the Naive Bayes classifier. The 20 newsgroups dataset has 20 classes and about 130,000 features per instance. Recall that the Naive Bayes classifer calculates a posterior distribution for each possible class. Each posterior distribution is a multiplication of many conditional distributions: 

$${\arg \max}_{j} \left(p(class=j)\prod_{i} p(attribute=i|class=j) \right)$$

where $p(class=j)$ is the prior and $p(attribute=i|class=j)$ is the conditional probability.

**Question for part 5**: With this many classes and features, how many different conditional probabilities need to be parameterized? How many priors need to be parameterized?

#### Answer part 5
20 priors need to be parameterized, with 20 x 130107 = 2,602,140 conditional probabilities.


___
## Naive Bayes in Scikit-learn
Scikit has several implementations of the Naive Bayes classifier: `GaussianNB`, `MultinomialNB`, and `BernoulliNB`. Look at the documentation here: http://scikit-learn.org/stable/modules/naive_bayes.html Take a look at each implementation and then answer this question: 

**Questions for part 6**: 
- If the instances contain mostly continuous attributes, would it be better to use Gaussian Naive Bayes, Multinomial Naive Bayes, or Bernoulli? And Why? 
- What if the data is sparse, does this change your answer? Why or Why not?

#### Answer part 6
For instances that are mostly continous attributres it is best to use Gaussian Naive Bayes because Multionoial and Bernoulli are more subitable for descrete attributes.

For sparse data we would have many absence of a feature it is best to use Bernoulli because it penalizes the feature.

## Naive Bayes Comparison
For the final section of this notebook let's compare the performance of Naive Bayes for document classification. Look at the parameters for `MultinomialNB`, and `BernoulliNB` (especially `alpha` and `binarize`). 

**Exercise for part 7**: Using the example code below, change the parameters for each classifier and see how accurate you can make the classifiers on the test set. 

**Question for part 7**: Why are these implementations so fast to train? What does the `'alpha'` value control in these models (*i.e.*, how does it change the parameterizations)? 

In [20]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB


clf_mnb = MultinomialNB(alpha=1.0)
clf_bnb = BernoulliNB(alpha=1.0, binarize=0.0)

results_bnb = []
results_mnb = []
for i in range(10):

    alpha = 0 +i*0.1
     
    clf_bnb = BernoulliNB(alpha=alpha, binarize=0.0)
    clf_bnb.fit(X_train, y_train)
    y_pred = clf_bnb.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print('BNB Percentage Accuracy ->', format(accuracy, ".3f"), 'alpha:', alpha)
    results_bnb.append((accuracy,alpha))
    
    
    
    clf_mnb = MultinomialNB(alpha=alpha)
    clf_mnb.fit(X_train, y_train)
    y_pred = clf_mnb.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print('MNB Percentage Accuracy ->', format(accuracy, ".3f"), 'alpha:', alpha)
    results_mnb.append((accuracy,alpha))

    print("\n")


C:\Datascience\Anaconda3\envs\ML7331\lib\site-packages\sklearn\naive_bayes.py:557: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  % _ALPHA_MIN


BNB Percentage Accuracy -> 0.852 alpha: 0.0
MNB Percentage Accuracy -> 0.855 alpha: 0.0




C:\Datascience\Anaconda3\envs\ML7331\lib\site-packages\sklearn\naive_bayes.py:557: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  % _ALPHA_MIN


BNB Percentage Accuracy -> 0.788 alpha: 0.1
MNB Percentage Accuracy -> 0.828 alpha: 0.1


BNB Percentage Accuracy -> 0.762 alpha: 0.2
MNB Percentage Accuracy -> 0.803 alpha: 0.2


BNB Percentage Accuracy -> 0.741 alpha: 0.30000000000000004
MNB Percentage Accuracy -> 0.782 alpha: 0.30000000000000004


BNB Percentage Accuracy -> 0.722 alpha: 0.4
MNB Percentage Accuracy -> 0.766 alpha: 0.4


BNB Percentage Accuracy -> 0.702 alpha: 0.5
MNB Percentage Accuracy -> 0.752 alpha: 0.5


BNB Percentage Accuracy -> 0.683 alpha: 0.6000000000000001
MNB Percentage Accuracy -> 0.740 alpha: 0.6000000000000001


BNB Percentage Accuracy -> 0.665 alpha: 0.7000000000000001
MNB Percentage Accuracy -> 0.728 alpha: 0.7000000000000001


BNB Percentage Accuracy -> 0.652 alpha: 0.8
MNB Percentage Accuracy -> 0.716 alpha: 0.8


BNB Percentage Accuracy -> 0.639 alpha: 0.9
MNB Percentage Accuracy -> 0.704 alpha: 0.9




These classifiers are so fast because these classifiers are so fast because there is no coefficient being optimized, just the probability for each class given different attribute values.

The alpha values controls how new occurrences in testing sets are handle, so that 0 probability is not present. The higher alpha is, the lower the probability.

________________________________________________________________________________________________________

That's all! Please **upload your rendered notebook to blackboard** and please include **team member names** in the notebook submission.